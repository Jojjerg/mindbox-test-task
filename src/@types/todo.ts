export interface ITodo {
  userId: number;
  id: number;
  title: string;
  completed: boolean;
}

export type TodoContextType = {
  todos: ITodo[];
  filter: IFilter;
  getTodos: () => void;
  addTodo: (title: string) => void;
  updateTodo: (id: number) => void;
  deleteCompletedTodos: () => void;
  changeFilter: (filter: IFilter) => void;
};

export type IFilter = "All" | "Completed" | "Active";
