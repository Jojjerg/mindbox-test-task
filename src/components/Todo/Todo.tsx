import CompletedSign from "../CompletedSign/CompletedSign";
import type { ITodo } from "../../@types/todo";
import React from "react";

interface TodoProps extends ITodo {
  updateTodo: (id: number) => void;
}

const Todo: React.FC<TodoProps> = ({ id, title, completed, updateTodo }) => {
  return (
    <div
      data-testid="todo"
      className="flex items-center justify-between gap-8 w-full px-4 py-3"
    >
      <div className="flex items-center gap-4">
        <div className="w-max h-max" style={{ opacity: completed ? 1 : 0 }}>
          <CompletedSign />
        </div>
        <div className="flex items-center justify-between">
          <p
            data-testid="title-text"
            className="text-lg"
            style={{
              textDecorationLine: completed ? "line-through" : "none",
              color: completed ? "#9ca3af" : "white",
            }}
          >
            {title}
          </p>
        </div>
      </div>
      <div>
        <button
          data-testid="update-button"
          className="px-2 py-1 rounded-md hover:brightness-[85%] transition-all duration-200"
          style={{
            background: completed ? "#ef4444" : "#32BEA6",
          }}
          onClick={() => updateTodo(id)}
        >
          <span data-testid="button-text" className="text-sm">
            {completed ? "Отменить" : "Выполнить"}
          </span>
        </button>
      </div>
    </div>
  );
};

export default Todo;
