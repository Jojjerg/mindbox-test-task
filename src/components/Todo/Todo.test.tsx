import { fireEvent, render, screen } from "@testing-library/react";

import { ITodo } from "../../@types/todo";
import React from "react";
import Todo from "./Todo";
import { act } from "react-test-renderer";
import { createRenderer } from "react-test-renderer/shallow";

const renderer = createRenderer();

const todos = [
  {
    userId: 1,
    id: 1,
    title: "delectus aut autem",
    completed: false,
  },
  {
    userId: 1,
    id: 2,
    title: "quis ut nam facilis et officia qui",
    completed: true,
  },
];

const updateTodo = (id: number) => {
  todos.filter((todo: ITodo) => {
    if (todo.id === id) {
      if (todo.completed) {
        todo.completed = false;
      } else {
        todo.completed = true;
      }
      return [...todos];
    }
  });
};

const defaultComponent = (
  <Todo
    userId={todos[1].userId}
    id={todos[1].id}
    title={todos[1].title}
    completed={todos[1].completed}
    updateTodo={updateTodo}
  />
);

describe("<Todo />", () => {
  it("should render and match the snapshot", () => {
    renderer.render(defaultComponent);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it("completed-sign renders correctly", () => {
    const { getByTestId } = render(defaultComponent);
    expect(getByTestId("completed-sign")).toBeInTheDocument();
  });

  it("title text equal to", () => {
    const { getByTestId } = render(defaultComponent);
    expect(getByTestId("title-text")).toHaveTextContent(
      "quis ut nam facilis et officia qui"
    );
  });

  it("button text equal to", () => {
    const { getByTestId } = render(defaultComponent);
    expect(getByTestId("button-text")).toHaveTextContent("Отменить");
  });

  it("update button works correctly", async () => {
    const { getByTestId } = render(defaultComponent);
    const updateButton = getByTestId("update-button");

    expect(updateButton).toBeVisible();

    await act(async () => {
      fireEvent.click(updateButton);
    });
  });
});
