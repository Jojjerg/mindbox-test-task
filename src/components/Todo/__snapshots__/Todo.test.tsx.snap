// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo /> should render and match the snapshot 1`] = `
<div
  className="flex items-center justify-between gap-8 w-full px-4 py-3"
  data-testid="todo"
>
  <div
    className="flex items-center gap-4"
  >
    <div
      className="w-max h-max"
      style={
        {
          "opacity": 1,
        }
      }
    >
      <CompletedSign />
    </div>
    <div
      className="flex items-center justify-between"
    >
      <p
        className="text-lg"
        data-testid="title-text"
        style={
          {
            "color": "#9ca3af",
            "textDecorationLine": "line-through",
          }
        }
      >
        quis ut nam facilis et officia qui
      </p>
    </div>
  </div>
  <div>
    <button
      className="px-2 py-1 rounded-md hover:brightness-[85%] transition-all duration-200"
      data-testid="update-button"
      onClick={[Function]}
      style={
        {
          "background": "#ef4444",
        }
      }
    >
      <span
        className="text-sm"
        data-testid="button-text"
      >
        Отменить
      </span>
    </button>
  </div>
</div>
`;
